### PR TITLE
feature: add volume filter

### PR DESCRIPTION
**Pull request by ZYecho**
Signed-off-by: zhangyue <zy675793960@yeah.net>

### Ⅰ. Describe what this PR did
Add `pouch volume ls` filter flag support.Support following filter type:
* driver, use driver name to do accurate match
* name, use volume name to do accurate match
* label, use label=key or label=key=value to do volume label filter

### Ⅱ. Does this pull request fix one issue?

#2462 

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
Add both api test case and cli test case.

### Ⅳ. Describe how to verify it
wait for CI.

### Ⅴ. Special notes for reviews
None.

### CODE REVIEW

File: storage/volume/core.go [hunk 1]
```diff
@@ -272,14 +274,25 @@ func (c *Core) ListVolumes(labels map[string]string) ([]*types.Volume, error) {
 		retVolumes = append(retVolumes, v)
 	}
 
-	// filter volumes if specify labels
-	if len(labels) != 0 {
+	// filter volumes
+	if filter.Len() > 0 {
```
Comment (HusterWan) [inline storage/volume/core.go]: How about 
```
if filter.Len() == 0 {
    return retVolumes, nil
}
```

we can decrease the indent, WDYT?

File: apis/filters/parse.go [hunk 2]
```diff
@@ -187,3 +187,32 @@ func (args Args) Validate(accepted map[string]bool) error {
 func FamiliarMatch(pattern string, ref string) (bool, error) {
 	return path.Match(pattern, ref)
 }
+
+// MatchKVList returns true if all the pairs in sources exist as key=value
+// pairs in the mapping at key, or if there are no values at key.
+func (args Args) MatchKVList(key string, sources map[string]string) bool {
```
Comment (zhuangqh) [inline apis/filters/parse.go]: This part of code is similar with https://github.com/alibaba/pouch/blob/master/daemon/mgr/container_list.go#L74
Is there any plan to unify them?

File: client/volume_list_test.go [hunk 3]
```diff
@@ -12,14 +12,15 @@ import (
 
 	"github.com/alibaba/pouch/apis/types"
 
+	"github.com/alibaba/pouch/apis/filters"
```
Comment (fuweid) [inline client/volume_list_test.go]: we could add the package into second group

File: apis/filters/parse.go (hunk 2 above)
Comment (ZYecho) [reply apis/filters/parse.go]: Yes, there's a plan, need to refactor to support a common filter package, as #2405.

File: test/cli_volume_test.go [hunk 4]
```diff
@@ -327,4 +327,44 @@ func (suite *PouchVolumeSuite) TestVolumeListOptions(c *check.C) {
...
+
+	// test filter options
+	volumeName4 := "volume_" + funcname + "4"
+	command.PouchRun("volume", "create", "--name", volumeName4, "--driver", "local", "--label", "test=foo").Assert(c, icmd.Success)
+	defer command.PouchRun("volume", "rm", volumeName4)
+	// test name filter
+	res := command.PouchRun("volume", "list", "--filter", "name="+volumeName4)
+	res.Assert(c, icmd.Success)
+
+	found := 0
```
Comment (fuweid) [inline test/cli_volume_test.go]: could we reorder the cases here? for example, use test case table to do that. I think there are a lot of duplicate code here. thanks.

File: storage/volume/core.go [hunk 5]
```diff
@@ -272,29 +274,40 @@ func (c *Core) ListVolumes(labels map[string]string) ([]*types.Volume, error) {
...
-		}
+	// filter volumes
+	if filter.Len() == 0 {
+		return retVolumes, nil
+	}
 
-		return filteredVolumes, nil
+	filteredVolumes := make([]*types.Volume, 0)
+	for _, vol := range retVolumes {
+		// do label filter
```
Comment (fuweid) [inline storage/volume/core.go]: just for confirm, filter only support `and` right?

Comment (ZYecho) [reply storage/volume/core.go]: relationship between different filter was `and`, but if multi same filter was use, such as `--driver=A, --driver=B`, there will be A or B.

Comment (fuweid) [reply storage/volume/core.go]: good to know. Thanks

File: test/cli_volume_test.go (hunk 4 above)
Comment (ZYecho) [reply test/cli_volume_test.go]: thanks for remind me, update.

File: test/cli_volume_test.go [hunk 6]
```diff
@@ -350,4 +350,29 @@ func (suite *PouchVolumeSuite) TestVolumeListOptions(c *check.C) {
...
+	// test filter options
+	volumeName4 := "volume_" + funcname + "4"
+	command.PouchRun("volume", "create", "--name", volumeName4, "--driver", "tmpfs", "--label", "test=foo").Assert(c, icmd.Success)
+	defer command.PouchRun("volume", "rm", volumeName4)
+
+	// test name, label, driver filter
+	filterArgs := []string{
+		"name=" + volumeName4,
+		"label=test",
+		"driver=tmpfs",
```
Comment (Ace-Tang) [inline test/cli_volume_test.go]: how about add some `and` case, like `--lable=test --name=a --driver=tmpfs`

Comment (ZYecho) [reply test/cli_volume_test.go]: Agree.

**Comment by HusterWan** [comment]
LGTM

**Pull request merged**